File: roll-your-own-test-framework/csharp/src/RollYourOwnTestFramework/Assertions.cs
namespace RollYourOwnTestFramework;

public static class Assertions
{
    public static void AssertEqual(object expected, object actual)
    {
        if (!Equals(expected, actual))
            throw new AssertionFailedException($"expected {expected} but got {actual}");
    }

    public static void AssertTrue(bool condition)
    {
        if (!condition)
            throw new AssertionFailedException("expected true");
    }

    public static void AssertThrows(Action action)
    {
        try
        {
            action();
        }
        catch
        {
            return;
        }

        throw new AssertionFailedException("expected exception");
    }
}

File: roll-your-own-test-framework/typescript/src/Assertions.ts
import { AssertionFailedException } from './AssertionFailedException.js';

export function assertEqual(expected: unknown, actual: unknown): void {
  if (expected !== actual) {
    throw new AssertionFailedException(`expected ${String(expected)} but got ${String(actual)}`);
  }
}

export function assertTrue(condition: boolean): void {
  if (!condition) {
    throw new AssertionFailedException('expected true');
  }
}

export function assertThrows(fn: () => void): void {
  try {
    fn();
  } catch {
    return;
  }
  throw new AssertionFailedException('expected exception');
}
